{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0e331693",
   "metadata": {},
   "source": [
    "# Test HAIL table creation from JSON files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "84675637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"edfa8101-3a73-4f7a-a999-401c625f4317\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "'use strict';\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    function drop(id) {\n",
       "      const view = Bokeh.index.get_by_id(id)\n",
       "      if (view != null) {\n",
       "        view.model.document.clear()\n",
       "        Bokeh.index.delete(view)\n",
       "      }\n",
       "    }\n",
       "\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null) {\n",
       "      drop(id)\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim()\n",
       "            drop(id)\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded(error = null) {\n",
       "    const el = document.getElementById(\"edfa8101-3a73-4f7a-a999-401c625f4317\");\n",
       "    if (el != null) {\n",
       "      const html = (() => {\n",
       "        if (typeof root.Bokeh === \"undefined\") {\n",
       "          if (error == null) {\n",
       "            return \"BokehJS is loading ...\";\n",
       "          } else {\n",
       "            return \"BokehJS failed to load.\";\n",
       "          }\n",
       "        } else {\n",
       "          const prefix = `BokehJS ${root.Bokeh.version}`;\n",
       "          if (error == null) {\n",
       "            return `${prefix} successfully loaded.`;\n",
       "          } else {\n",
       "            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n",
       "          }\n",
       "        }\n",
       "      })();\n",
       "      el.innerHTML = html;\n",
       "\n",
       "      if (error != null) {\n",
       "        const wrapper = document.createElement(\"div\");\n",
       "        wrapper.style.overflow = \"auto\";\n",
       "        wrapper.style.height = \"5em\";\n",
       "        wrapper.style.resize = \"vertical\";\n",
       "        const content = document.createElement(\"div\");\n",
       "        content.style.fontFamily = \"monospace\";\n",
       "        content.style.whiteSpace = \"pre-wrap\";\n",
       "        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n",
       "        content.textContent = error.stack ?? error.toString();\n",
       "        wrapper.append(content);\n",
       "        el.append(wrapper);\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(() => display_loaded(error), 100);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.4.3.min.js\"];\n",
       "  const css_urls = [];\n",
       "\n",
       "  const inline_js = [    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "function(Bokeh) {\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      try {\n",
       "            for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "\n",
       "      } catch (error) {display_loaded(error);throw error;\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"edfa8101-3a73-4f7a-a999-401c625f4317\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "'use strict';\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded(error = null) {\n    const el = document.getElementById(\"edfa8101-3a73-4f7a-a999-401c625f4317\");\n    if (el != null) {\n      const html = (() => {\n        if (typeof root.Bokeh === \"undefined\") {\n          if (error == null) {\n            return \"BokehJS is loading ...\";\n          } else {\n            return \"BokehJS failed to load.\";\n          }\n        } else {\n          const prefix = `BokehJS ${root.Bokeh.version}`;\n          if (error == null) {\n            return `${prefix} successfully loaded.`;\n          } else {\n            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n          }\n        }\n      })();\n      el.innerHTML = html;\n\n      if (error != null) {\n        const wrapper = document.createElement(\"div\");\n        wrapper.style.overflow = \"auto\";\n        wrapper.style.height = \"5em\";\n        wrapper.style.resize = \"vertical\";\n        const content = document.createElement(\"div\");\n        content.style.fontFamily = \"monospace\";\n        content.style.whiteSpace = \"pre-wrap\";\n        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n        content.textContent = error.stack ?? error.toString();\n        wrapper.append(content);\n        el.append(wrapper);\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(() => display_loaded(error), 100);\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n      try {\n            for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n\n      } catch (error) {display_loaded(error);throw error;\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"edfa8101-3a73-4f7a-a999-401c625f4317\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/11/17 12:06:25 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "Running on Apache Spark version 3.5.7\n",
      "SparkUI available at http://r860:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.136-c32f88309ab0\n",
      "LOGGING: writing to /dados/bioinfo.time/repos/gnomad-fc/code/hail-20251117-1206-0.2.136-c32f88309ab0.log\n"
     ]
    }
   ],
   "source": [
    "import hail as hl\n",
    "import pandas as pd\n",
    "\n",
    "hl.init(spark_conf={\n",
    "        'spark.driver.memory': '320g',\n",
    "        'spark.executor.memory': '320g',\n",
    "        'spark.driver.maxResultSize': '100g',\n",
    "        'spark.kryoserializer.buffer.max': '2047G'\n",
    "    })\n",
    "\n",
    "# Load your table\n",
    "ht = hl.read_table(\"/dados/bioinfo.time/repos/gnomad-fc/data/variants.exo.ht/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "addea338",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Quick checks\n",
    "# print(f\"Total variants: {ht.count()}\")\n",
    "# print(f\"Chromosomes: {ht.aggregate(hl.agg.collect_as_set(ht.chromosome))}\")\n",
    "# print(f\"Fields: {list(ht.row)}\")\n",
    "\n",
    "# # Look at first few rows\n",
    "# ht.show(5)\n",
    "\n",
    "# # Get summary statistics\n",
    "# ht.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "985224bf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------\n",
      "Global fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'chromosome': str \n",
      "    'position': int32 \n",
      "    'ref': str \n",
      "    'alt': str \n",
      "    'vid': str \n",
      "    'hgvsg': str \n",
      "    'variant_type': str \n",
      "    'begin': int32 \n",
      "    'end': int32 \n",
      "    'filters': str \n",
      "    'mapping_quality': float64 \n",
      "    'fisher_strand_bias': float64 \n",
      "    'quality': float64 \n",
      "    'cytogenetic_band': str \n",
      "    'phylop_score': float64 \n",
      "    'phylop_primate_score': float64 \n",
      "    'gerp_score': float64 \n",
      "    'dann_score': float64 \n",
      "    'rsid': str \n",
      "    'gnomad_af': float64 \n",
      "    'gnomad_ac': int32 \n",
      "    'gnomad_an': int32 \n",
      "    'gnomad_hc': int32 \n",
      "    'gnomad_afr_af': float64 \n",
      "    'gnomad_amr_af': float64 \n",
      "    'gnomad_eas_af': float64 \n",
      "    'gnomad_fin_af': float64 \n",
      "    'gnomad_nfe_af': float64 \n",
      "    'gnomad_asj_af': float64 \n",
      "    'gnomad_sas_af': float64 \n",
      "    'gnomad_oth_af': float64 \n",
      "    'gnomad_failed_filter': bool \n",
      "    'gnomad_exome_af': float64 \n",
      "    'gnomad_exome_ac': int32 \n",
      "    'gnomad_exome_an': int32 \n",
      "    'gnomad_exome_hc': int32 \n",
      "    'gnomad_exome_failed_filter': bool \n",
      "    'topmed_af': float64 \n",
      "    'topmed_ac': int32 \n",
      "    'topmed_an': int32 \n",
      "    'topmed_hc': int32 \n",
      "    'topmed_failed_filter': bool \n",
      "    'clinvar_significance': str \n",
      "    'clinvar_id': str \n",
      "    'n_transcripts': int32 \n",
      "    'transcripts': array<struct {\n",
      "        transcript_id: str, \n",
      "        source: str, \n",
      "        bio_type: str, \n",
      "        gene_id: str, \n",
      "        hgnc: str, \n",
      "        consequences: array<str>, \n",
      "        impact: str, \n",
      "        is_canonical: bool\n",
      "    }> \n",
      "    'samples': array<struct {\n",
      "        genotype: str, \n",
      "        variant_frequencies: array<float64>, \n",
      "        total_depth: int32, \n",
      "        genotype_quality: int32, \n",
      "        allele_depths: array<int32>\n",
      "    }> \n",
      "    'locus': locus<GRCh38> \n",
      "    'alleles': array<str> \n",
      "    'max_gnomad_af': float64 \n",
      "    'max_pop_af': float64 \n",
      "    'canonical_transcript.transcript_id': str \n",
      "    'canonical_transcript.source': str \n",
      "    'canonical_transcript.bio_type': str \n",
      "    'canonical_transcript.gene_id': str \n",
      "    'canonical_transcript.hgnc': str \n",
      "    'canonical_transcript.consequences': array<str> \n",
      "    'canonical_transcript.impact': str \n",
      "    'canonical_transcript.is_canonical': bool \n",
      "    'all_consequences': set<str> \n",
      "    'genes': set<str> \n",
      "----------------------------------------\n",
      "Key: []\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "ht = ht.flatten()\n",
    "ht.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0082a6e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Config Map(es.nodes.wan.only -> true, es.batch.size.entries -> 10000, es.index.auto.create -> true, es.port -> 9200, es.nodes -> 172.17.0.2)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 1:=================================================>(65466 + 64) / 65530]"
     ]
    }
   ],
   "source": [
    "# es_host = 'localhost'\n",
    "es_host = \"172.17.0.2\"\n",
    "es_port = 9200\n",
    "es_index = 'fiocruz_variants'\n",
    "hl.export_elasticsearch(\n",
    "    ht,\n",
    "    host=es_host,\n",
    "    port=es_port,\n",
    "    index=es_index,\n",
    "    index_type='_doc',\n",
    "    block_size=10000,\n",
    "    config={\n",
    "        \"es.nodes.wan.only\": \"true\"   # necessário quando acessar container via IP\n",
    "    }\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c181f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def explore_hail_table(ht: hl.Table):\n",
    "    \"\"\"\n",
    "    Comprehensive exploration and testing of a Hail Table\n",
    "    \n",
    "    Args:\n",
    "        ht (hl.Table): Hail Table to explore\n",
    "    \"\"\"\n",
    "    print(\"=\"*80)\n",
    "    print(\"HAIL TABLE EXPLORATION AND TESTING\")\n",
    "    print(\"=\"*80)\n",
    "    \n",
    "    # Basic info\n",
    "    print(\"\\n2. BASIC STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    n_variants = ht.count()\n",
    "    print(f\"   Total variants: {n_variants:,}\")\n",
    "    print(f\"   Key fields: {ht.key}\")\n",
    "    \n",
    "    # Schema\n",
    "    print(\"\\n3. SCHEMA OVERVIEW\")\n",
    "    print(\"-\" * 80)\n",
    "    ht.describe()\n",
    "    \n",
    "    # Field names\n",
    "    print(\"\\n4. ALL FIELD NAMES\")\n",
    "    print(\"-\" * 80)\n",
    "    fields = list(ht.row)\n",
    "    for i, field in enumerate(fields, 1):\n",
    "        print(f\"   {i:2d}. {field}\")\n",
    "    \n",
    "    # Sample data - first 5 rows\n",
    "    print(\"\\n5. SAMPLE DATA (First 5 variants)\")\n",
    "    print(\"-\" * 80)\n",
    "    sample_df = ht.head(5).to_pandas()\n",
    "    print(sample_df[['chromosome', 'position', 'ref', 'alt', 'vid', 'rsid']].to_string())\n",
    "    \n",
    "    # Chromosome distribution\n",
    "    print(\"\\n6. CHROMOSOME DISTRIBUTION\")\n",
    "    print(\"-\" * 80)\n",
    "    chrom_counts = ht.group_by(ht.chromosome).aggregate(n=hl.agg.count()).collect()\n",
    "    chrom_df = pd.DataFrame(chrom_counts).sort_values('chromosome')\n",
    "    print(chrom_df.to_string(index=False))\n",
    "    \n",
    "    # Variant type distribution\n",
    "    print(\"\\n7. VARIANT TYPE DISTRIBUTION\")\n",
    "    print(\"-\" * 80)\n",
    "    var_types = ht.group_by(ht.variant_type).aggregate(n=hl.agg.count()).collect()\n",
    "    var_type_df = pd.DataFrame(var_types).sort_values('n', ascending=False)\n",
    "    print(var_type_df.to_string(index=False))\n",
    "    \n",
    "    # gnomAD frequency statistics\n",
    "    print(\"\\n8. GNOMAD ALLELE FREQUENCY STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    gnomad_stats = ht.aggregate(hl.struct(\n",
    "        total_variants=hl.agg.count(),\n",
    "        with_gnomad_af=hl.agg.count_where(hl.is_defined(ht.gnomad_af)),\n",
    "        with_gnomad_exome_af=hl.agg.count_where(hl.is_defined(ht.gnomad_exome_af)),\n",
    "        mean_gnomad_af=hl.agg.mean(ht.gnomad_af),\n",
    "        median_gnomad_af=hl.agg.approx_median(ht.gnomad_af),\n",
    "        max_gnomad_af=hl.agg.max(ht.gnomad_af),\n",
    "    ))\n",
    "    print(f\"   Total variants: {gnomad_stats.total_variants:,}\")\n",
    "    print(f\"   With gnomAD genome AF: {gnomad_stats.with_gnomad_af:,}\")\n",
    "    print(f\"   With gnomAD exome AF: {gnomad_stats.with_gnomad_exome_af:,}\")\n",
    "    print(f\"   Mean gnomAD AF: {gnomad_stats.mean_gnomad_af:.6f}\" if gnomad_stats.mean_gnomad_af else \"   Mean gnomAD AF: N/A\")\n",
    "    print(f\"   Median gnomAD AF: {gnomad_stats.median_gnomad_af:.6f}\" if gnomad_stats.median_gnomad_af else \"   Median gnomAD AF: N/A\")\n",
    "    print(f\"   Max gnomAD AF: {gnomad_stats.max_gnomad_af:.6f}\" if gnomad_stats.max_gnomad_af else \"   Max gnomAD AF: N/A\")\n",
    "    \n",
    "    # Frequency bins\n",
    "    print(\"\\n9. GNOMAD AF FREQUENCY BINS\")\n",
    "    print(\"-\" * 80)\n",
    "    freq_bins = ht.aggregate(hl.struct(\n",
    "        rare_0_0001=hl.agg.count_where((ht.max_gnomad_af > 0) & (ht.max_gnomad_af <= 0.0001)),\n",
    "        rare_0_001=hl.agg.count_where((ht.max_gnomad_af > 0.0001) & (ht.max_gnomad_af <= 0.001)),\n",
    "        low_freq_0_01=hl.agg.count_where((ht.max_gnomad_af > 0.001) & (ht.max_gnomad_af <= 0.01)),\n",
    "        common_0_05=hl.agg.count_where((ht.max_gnomad_af > 0.01) & (ht.max_gnomad_af <= 0.05)),\n",
    "        very_common=hl.agg.count_where(ht.max_gnomad_af > 0.05),\n",
    "        no_data=hl.agg.count_where(~hl.is_defined(ht.max_gnomad_af)),\n",
    "    ))\n",
    "    print(f\"   Ultra-rare (AF ≤ 0.0001): {freq_bins.rare_0_0001:,}\")\n",
    "    print(f\"   Rare (0.0001 < AF ≤ 0.001): {freq_bins.rare_0_001:,}\")\n",
    "    print(f\"   Low freq (0.001 < AF ≤ 0.01): {freq_bins.low_freq_0_01:,}\")\n",
    "    print(f\"   Common (0.01 < AF ≤ 0.05): {freq_bins.common_0_05:,}\")\n",
    "    print(f\"   Very common (AF > 0.05): {freq_bins.very_common:,}\")\n",
    "    print(f\"   No gnomAD data: {freq_bins.no_data:,}\")\n",
    "    \n",
    "    # Transcript statistics\n",
    "    print(\"\\n10. TRANSCRIPT STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    transcript_stats = ht.aggregate(hl.struct(\n",
    "        with_transcripts=hl.agg.count_where(ht.n_transcripts > 0),\n",
    "        total_transcripts=hl.agg.sum(ht.n_transcripts),\n",
    "        with_canonical=hl.agg.count_where(hl.is_defined(ht.canonical_transcript)),\n",
    "        mean_transcripts=hl.agg.mean(ht.n_transcripts),\n",
    "        max_transcripts=hl.agg.max(ht.n_transcripts),\n",
    "    ))\n",
    "    print(f\"   Variants with transcripts: {transcript_stats.with_transcripts:,}\")\n",
    "    print(f\"   Total transcript annotations: {transcript_stats.total_transcripts:,}\")\n",
    "    print(f\"   Variants with canonical transcript: {transcript_stats.with_canonical:,}\")\n",
    "    print(f\"   Mean transcripts per variant: {transcript_stats.mean_transcripts:.2f}\")\n",
    "    print(f\"   Max transcripts per variant: {transcript_stats.max_transcripts}\")\n",
    "    \n",
    "    # ClinVar statistics\n",
    "    print(\"\\n11. CLINVAR ANNOTATIONS\")\n",
    "    print(\"-\" * 80)\n",
    "    clinvar_stats = ht.aggregate(hl.struct(\n",
    "        with_clinvar=hl.agg.count_where(hl.is_defined(ht.clinvar_significance)),\n",
    "        total=hl.agg.count(),\n",
    "    ))\n",
    "    print(f\"   Variants with ClinVar: {clinvar_stats.with_clinvar:,}\")\n",
    "    print(f\"   Percentage: {100 * clinvar_stats.with_clinvar / clinvar_stats.total:.2f}%\")\n",
    "    \n",
    "    # Conservation scores\n",
    "    print(\"\\n12. CONSERVATION SCORE STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    conservation_stats = ht.aggregate(hl.struct(\n",
    "        phylop_mean=hl.agg.mean(ht.phylop_score),\n",
    "        phylop_max=hl.agg.max(ht.phylop_score),\n",
    "        gerp_mean=hl.agg.mean(ht.gerp_score),\n",
    "        gerp_max=hl.agg.max(ht.gerp_score),\n",
    "    ))\n",
    "    print(f\"   PhyloP mean: {conservation_stats.phylop_mean:.3f}\" if conservation_stats.phylop_mean else \"   PhyloP mean: N/A\")\n",
    "    print(f\"   PhyloP max: {conservation_stats.phylop_max:.3f}\" if conservation_stats.phylop_max else \"   PhyloP max: N/A\")\n",
    "    print(f\"   GERP mean: {conservation_stats.gerp_mean:.3f}\" if conservation_stats.gerp_mean else \"   GERP mean: N/A\")\n",
    "    print(f\"   GERP max: {conservation_stats.gerp_max:.3f}\" if conservation_stats.gerp_max else \"   GERP max: N/A\")\n",
    "    \n",
    "    # Gene statistics\n",
    "    print(\"\\n13. GENE STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    gene_stats = ht.aggregate(hl.struct(\n",
    "        variants_with_genes=hl.agg.count_where(hl.len(ht.genes) > 0),\n",
    "    ))\n",
    "    # Count unique genes by converting set to array and aggregating\n",
    "    unique_genes_count = ht.aggregate(\n",
    "        hl.len(hl.agg.explode(lambda gene: hl.agg.collect_as_set(gene), hl.array(ht.genes)))\n",
    "    )\n",
    "    print(f\"   Variants with gene annotations: {gene_stats.variants_with_genes:,}\")\n",
    "    print(f\"   Unique genes: {unique_genes_count:,}\")\n",
    "    \n",
    "    # Top consequences\n",
    "    print(\"\\n14. TOP 10 CONSEQUENCES\")\n",
    "    print(\"-\" * 80)\n",
    "    consequences_ht = ht.explode(ht.all_consequences)\n",
    "    top_consequences = (consequences_ht\n",
    "                       .group_by(consequences_ht.all_consequences)\n",
    "                       .aggregate(n=hl.agg.count())\n",
    "                       .order_by(hl.desc('n'))\n",
    "                       .head(10)\n",
    "                       .collect())\n",
    "    for i, row in enumerate(top_consequences, 1):\n",
    "        print(f\"   {i:2d}. {row.all_consequences}: {row.n:,}\")\n",
    "    \n",
    "    # Example queries\n",
    "    print(\"\\n15. EXAMPLE VARIANT LOOKUPS\")\n",
    "    print(\"-\" * 80)\n",
    "    \n",
    "    # Get a few interesting variants\n",
    "    # Common variant\n",
    "    common = ht.filter(ht.max_gnomad_af > 0.1).head(1).collect()\n",
    "    if common:\n",
    "        v = common[0]\n",
    "        print(f\"\\n   Common variant (AF > 0.1):\")\n",
    "        print(f\"   {v.chromosome}:{v.position} {v.ref}>{v.alt}\")\n",
    "        print(f\"   rsID: {v.rsid}\")\n",
    "        print(f\"   gnomAD AF: {v.gnomad_af:.4f}\" if v.gnomad_af else \"   gnomAD AF: N/A\")\n",
    "        print(f\"   Genes: {', '.join(v.genes) if v.genes else 'N/A'}\")\n",
    "    \n",
    "    # Rare variant\n",
    "    rare = ht.filter((ht.max_gnomad_af > 0) & (ht.max_gnomad_af < 0.001)).head(1).collect()\n",
    "    if rare:\n",
    "        v = rare[0]\n",
    "        print(f\"\\n   Rare variant (AF < 0.001):\")\n",
    "        print(f\"   {v.chromosome}:{v.position} {v.ref}>{v.alt}\")\n",
    "        print(f\"   gnomAD AF: {v.gnomad_af:.6f}\" if v.gnomad_af else \"   gnomAD AF: N/A\")\n",
    "        print(f\"   Consequences: {', '.join(list(v.all_consequences)) if v.all_consequences else 'N/A'}\")\n",
    "    \n",
    "    # Variant with ClinVar\n",
    "    clinvar_var = ht.filter(hl.is_defined(ht.clinvar_significance)).head(1).collect()\n",
    "    if clinvar_var:\n",
    "        v = clinvar_var[0]\n",
    "        print(f\"\\n   ClinVar variant:\")\n",
    "        print(f\"   {v.chromosome}:{v.position} {v.ref}>{v.alt}\")\n",
    "        print(f\"   ClinVar significance: {v.clinvar_significance}\")\n",
    "        print(f\"   ClinVar ID: {v.clinvar_id}\")\n",
    "    \n",
    "    print(\"\\n\" + \"=\"*80)\n",
    "    print(\"TESTING COMPLETE!\")\n",
    "    print(\"=\"*80)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e4d4ee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "explore_hail_table(ht)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4405be8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_examples(ht: hl.Table):\n",
    "    \"\"\"\n",
    "    Examples of useful queries you can run\n",
    "    \n",
    "    Args:\n",
    "        ht (hl.Table): Hail Table to query\n",
    "    \"\"\"\n",
    "    print(\"\\n\" + \"=\"*80)\n",
    "    print(\"EXAMPLE QUERIES\")\n",
    "    print(\"=\"*80)\n",
    "    \n",
    "    # 1. Filter by frequency\n",
    "    print(\"\\n1. Get rare variants (AF < 0.01)\")\n",
    "    rare_variants = ht.filter(\n",
    "        (ht.max_gnomad_af > 0) & (ht.max_gnomad_af < 0.01)\n",
    "    )\n",
    "    print(f\"   Found {rare_variants.count():,} rare variants\")\n",
    "    \n",
    "    # 2. Filter by gene\n",
    "    print(\"\\n2. Get variants in specific gene (e.g., first gene in dataset)\")\n",
    "    first_gene = ht.filter(hl.len(ht.genes) > 0).head(1).collect()[0].genes\n",
    "    if first_gene:\n",
    "        gene_name = list(first_gene)[0]\n",
    "        gene_variants = ht.filter(ht.genes.contains(gene_name))\n",
    "        print(f\"   Gene: {gene_name}\")\n",
    "        print(f\"   Variants: {gene_variants.count()}\")\n",
    "    \n",
    "    # 3. Filter by consequence\n",
    "    print(\"\\n3. Get missense variants\")\n",
    "    missense = ht.filter(ht.all_consequences.contains('missense_variant'))\n",
    "    print(f\"   Found {missense.count():,} missense variants\")\n",
    "    \n",
    "    # 4. Filter by ClinVar pathogenic\n",
    "    print(\"\\n4. Get pathogenic variants\")\n",
    "    pathogenic = ht.filter(\n",
    "        hl.is_defined(ht.clinvar_significance) &\n",
    "        ht.clinvar_significance.contains('Pathogenic')\n",
    "    )\n",
    "    print(f\"   Found {pathogenic.count():,} pathogenic variants\")\n",
    "    \n",
    "    # 5. Filter by conservation\n",
    "    print(\"\\n5. Get highly conserved variants (PhyloP > 2)\")\n",
    "    conserved = ht.filter(ht.phylop_score > 2)\n",
    "    print(f\"   Found {conserved.count():,} highly conserved variants\")\n",
    "    \n",
    "    # 6. Complex filter\n",
    "    print(\"\\n6. Complex filter: Rare + missense + conserved\")\n",
    "    complex_filter = ht.filter(\n",
    "        (ht.max_gnomad_af < 0.01) &\n",
    "        ht.all_consequences.contains('missense_variant') &\n",
    "        (ht.phylop_score > 2)\n",
    "    )\n",
    "    print(f\"   Found {complex_filter.count():,} variants\")\n",
    "    \n",
    "    print(\"\\n\" + \"=\"*80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efba17f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_examples(ht)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
