{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0e331693",
   "metadata": {},
   "source": [
    "# Test HAIL table creation from JSON files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84675637",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "    <style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"c9c95426-2740-4745-bb8f-1566b302eb67\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "'use strict';\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    function drop(id) {\n      const view = Bokeh.index.get_by_id(id)\n      if (view != null) {\n        view.model.document.clear()\n        Bokeh.index.delete(view)\n      }\n    }\n\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n\n    // Clean up Bokeh references\n    if (id != null) {\n      drop(id)\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim()\n            drop(id)\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded(error = null) {\n    const el = document.getElementById(\"c9c95426-2740-4745-bb8f-1566b302eb67\");\n    if (el != null) {\n      const html = (() => {\n        if (typeof root.Bokeh === \"undefined\") {\n          if (error == null) {\n            return \"BokehJS is loading ...\";\n          } else {\n            return \"BokehJS failed to load.\";\n          }\n        } else {\n          const prefix = `BokehJS ${root.Bokeh.version}`;\n          if (error == null) {\n            return `${prefix} successfully loaded.`;\n          } else {\n            return `${prefix} <b>encountered errors</b> while loading and may not function as expected.`;\n          }\n        }\n      })();\n      el.innerHTML = html;\n\n      if (error != null) {\n        const wrapper = document.createElement(\"div\");\n        wrapper.style.overflow = \"auto\";\n        wrapper.style.height = \"5em\";\n        wrapper.style.resize = \"vertical\";\n        const content = document.createElement(\"div\");\n        content.style.fontFamily = \"monospace\";\n        content.style.whiteSpace = \"pre-wrap\";\n        content.style.backgroundColor = \"rgb(255, 221, 221)\";\n        content.textContent = error.stack ?? error.toString();\n        wrapper.append(content);\n        el.append(wrapper);\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(() => display_loaded(error), 100);\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.4.3.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.4.3.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n      try {\n            for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n\n      } catch (error) {display_loaded(error);throw error;\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"c9c95426-2740-4745-bb8f-1566b302eb67\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/11/08 23:45:36 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "Running on Apache Spark version 3.5.7\n",
      "SparkUI available at http://pimenta:4040\n",
      "Welcome to\n",
      "     __  __     <>__\n",
      "    / /_/ /__  __/ /\n",
      "   / __  / _ `/ / /\n",
      "  /_/ /_/\\_,_/_/_/   version 0.2.136-c32f88309ab0\n",
      "LOGGING: writing to /home/matheus/Documents/gnomeAD-tb/code/hail-20251108-2345-0.2.136-c32f88309ab0.log\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-08 23:45:59.010 Hail: INFO: Ordering unsorted dataset with network shuffle\n",
      "2025-11-08 23:46:00.608 Hail: INFO: Ordering unsorted dataset with network shuffle\n",
      "2025-11-08 23:46:05.465 Hail: INFO: Ordering unsorted dataset with network shuffle\n"
     ]
    }
   ],
   "source": [
    "import hail as hl\n",
    "import pandas as pd\n",
    "hl.init()\n",
    "\n",
    "# Load your table\n",
    "ht = hl.read_table(\"../data/json_variants.ht\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "addea338",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total variants: 30487\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 0:==============================>                         (46 + 14) / 84]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chromosomes: {'chr2', 'chr21', 'chr3', 'chr12', 'chr14', 'chr7', 'chr22', 'chr10', 'chr9', 'chr11', 'chr5', 'chr13', 'chrX', 'chrY', 'chr19', 'chr18', 'chr6', 'chr4', 'chr17', 'chr15', 'chr20', 'chr8', 'chr16', 'chr1'}\n",
      "Fields: ['chromosome', 'position', 'ref', 'alt', 'vid', 'hgvsg', 'variant_type', 'begin', 'end', 'filters', 'mapping_quality', 'fisher_strand_bias', 'quality', 'cytogenetic_band', 'phylop_score', 'phylop_primate_score', 'gerp_score', 'dann_score', 'rsid', 'gnomad_af', 'gnomad_ac', 'gnomad_an', 'gnomad_hc', 'gnomad_afr_af', 'gnomad_amr_af', 'gnomad_eas_af', 'gnomad_fin_af', 'gnomad_nfe_af', 'gnomad_asj_af', 'gnomad_sas_af', 'gnomad_oth_af', 'gnomad_failed_filter', 'gnomad_exome_af', 'gnomad_exome_ac', 'gnomad_exome_an', 'gnomad_exome_hc', 'gnomad_exome_failed_filter', 'topmed_af', 'topmed_ac', 'topmed_an', 'topmed_hc', 'topmed_failed_filter', 'clinvar_significance', 'clinvar_id', 'n_transcripts', 'transcripts', 'samples', 'locus', 'alleles', 'max_gnomad_af', 'max_pop_af', 'canonical_transcript', 'all_consequences', 'genes']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table><thead><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"8\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"8\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">canonical_transcript</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;\"></div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">chromosome</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">position</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">ref</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">alt</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">vid</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">hgvsg</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">variant_type</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">begin</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">end</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">filters</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">mapping_quality</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">fisher_strand_bias</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">quality</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">cytogenetic_band</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">phylop_score</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">phylop_primate_score</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gerp_score</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">dann_score</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">rsid</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_ac</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_an</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_hc</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_afr_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_amr_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_eas_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_fin_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_nfe_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_asj_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_sas_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_oth_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_failed_filter</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_exome_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_exome_ac</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_exome_an</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_exome_hc</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gnomad_exome_failed_filter</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">topmed_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">topmed_ac</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">topmed_an</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">topmed_hc</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">topmed_failed_filter</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">clinvar_significance</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">clinvar_id</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">n_transcripts</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">transcripts</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">samples</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">locus</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">alleles</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">max_gnomad_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">max_pop_af</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">transcript_id</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">source</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">bio_type</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">gene_id</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">hgnc</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">consequences</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">impact</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">is_canonical</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">all_consequences</div></td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \" colspan=\"1\"><div style=\"text-align: left;border-bottom: solid 2px #000; padding-bottom: 5px\">genes</div></td></tr><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">bool</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">bool</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">bool</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">int32</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">array&lt;struct{transcript_id: str, source: str, bio_type: str, gene_id: str, hgnc: str, consequences: array&lt;str&gt;, impact: str, is_canonical: bool}&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">array&lt;struct{genotype: str, variant_frequencies: array&lt;float64&gt;, total_depth: int32, genotype_quality: int32, allele_depths: array&lt;int32&gt;}&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">locus&lt;GRCh38&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">array&lt;str&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">float64</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">array&lt;str&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">str</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">bool</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">set&lt;str&gt;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; text-align: left;\">set&lt;str&gt;</td></tr>\n",
       "</thead><tbody><tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;chr1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69081</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1-69081-G-C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;NC_000001.11:g.69081G&gt;C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;SNV&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69081</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69081</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;PASS&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,29e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1p36.33&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">-5,00e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">-1,65e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">-5,68e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3,60e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,50e-03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">186</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">21896</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,01e-03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,32e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">0,00e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3,10e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,12e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,27e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5,17e-03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,76e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4788</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">271898</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">729</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,29e-03</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1041</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">125568</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">0</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;ENST00000642116.1&quot;,&quot;Ensembl&quot;,&quot;lncRNA&quot;,&quot;ENSG00000290826&quot;,&quot;ENSG00000290826&quot;,[&quot;downstream_gene_variant&quot;],&quot;modifier&quot;,True),(&quot;ENST00000641515.2&quot;,&quot;Ensembl&quot;,&quot;mRNA&quot;,&quot;ENSG00000186092&quot;,&quot;OR4F5&quot;,[&quot;synonymous_variant&quot;],&quot;low&quot;,True),(&quot;NM_001005484.2&quot;,&quot;RefSeq&quot;,&quot;mRNA&quot;,&quot;79501&quot;,&quot;OR4F5&quot;,[&quot;synonymous_variant&quot;],&quot;low&quot;,True)]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;0/1&quot;,[2,85e-01],116,9,[83,33])]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:69081</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;G&quot;,&quot;C&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,76e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3,10e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENST00000642116.1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Ensembl&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;lncRNA&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000290826&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000290826&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;downstream_gene_variant&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;modifier&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;downstream_gene_variant&quot;,&quot;synonymous_variant&quot;}</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;ENSG00000290826&quot;,&quot;OR4F5&quot;}</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;chr1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69270</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;A&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1-69270-A-G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;NC_000001.11:g.69270A&gt;G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;SNV&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69270</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69270</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;PASS&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,22e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1p36.33&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,00e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,08e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">-2,89e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5,80e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,29e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">13469</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">21428</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4877</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">3,54e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,95e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,96e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,29e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,80e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,03e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,93e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,78e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">374631</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">426620</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">176537</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,75e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">34480</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">125568</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">11</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;ENST00000641515.2&quot;,&quot;Ensembl&quot;,&quot;mRNA&quot;,&quot;ENSG00000186092&quot;,&quot;OR4F5&quot;,[&quot;synonymous_variant&quot;],&quot;low&quot;,True),(&quot;NM_001005484.2&quot;,&quot;RefSeq&quot;,&quot;mRNA&quot;,&quot;79501&quot;,&quot;OR4F5&quot;,[&quot;synonymous_variant&quot;],&quot;low&quot;,True)]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;0/1&quot;,[6,75e-01],83,3,[27,56])]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:69270</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;A&quot;,&quot;G&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,78e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,96e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENST00000641515.2&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Ensembl&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;mRNA&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000186092&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;OR4F5&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;synonymous_variant&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;low&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;synonymous_variant&quot;}</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;OR4F5&quot;}</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;chr1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69511</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;A&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1-69511-A-G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;NC_000001.11:g.69511A&gt;G&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;SNV&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69511</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69511</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;PASS&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,98e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1p36.33&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">-1,00e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,99e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,29e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4,50e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,46e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">70558</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">83430</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">32520</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5,95e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,95e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,00e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,91e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,67e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,78e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,78e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,62e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">832303</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">865570</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">410006</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">8,75e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">109911</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">125568</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">50718</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;ENST00000641515.2&quot;,&quot;Ensembl&quot;,&quot;mRNA&quot;,&quot;ENSG00000186092&quot;,&quot;OR4F5&quot;,[&quot;missense_variant&quot;],&quot;moderate&quot;,True),(&quot;NM_001005484.2&quot;,&quot;RefSeq&quot;,&quot;mRNA&quot;,&quot;79501&quot;,&quot;OR4F5&quot;,[&quot;missense_variant&quot;],&quot;moderate&quot;,True)]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;1/1&quot;,[1,00e+00],68,41,[0,68])]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:69511</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;A&quot;,&quot;G&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,62e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,00e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENST00000641515.2&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Ensembl&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;mRNA&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000186092&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;OR4F5&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;missense_variant&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;moderate&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;missense_variant&quot;}</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;OR4F5&quot;}</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;chr1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69897</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;T&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1-69897-T-C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;NC_000001.11:g.69897T&gt;C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;SNV&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69897</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">69897</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;PASS&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,03e+01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1p36.33&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7,00e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">1,05e+00</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">9,04e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,20e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4,86e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">25259</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">51958</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">5492</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,92e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4,78e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,04e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,99e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,75e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,71e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,26e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7,04e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">209373</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">297420</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">78375</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,80e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">35215</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">125568</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">25</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;ENST00000641515.2&quot;,&quot;Ensembl&quot;,&quot;mRNA&quot;,&quot;ENSG00000186092&quot;,&quot;OR4F5&quot;,[&quot;synonymous_variant&quot;],&quot;low&quot;,True),(&quot;NM_001005484.2&quot;,&quot;RefSeq&quot;,&quot;mRNA&quot;,&quot;79501&quot;,&quot;OR4F5&quot;,[&quot;synonymous_variant&quot;],&quot;low&quot;,True)]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;0/1&quot;,[6,67e-01],102,6,[34,68])]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:69897</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;T&quot;,&quot;C&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">7,04e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,99e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENST00000641515.2&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Ensembl&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;mRNA&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000186092&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;OR4F5&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;synonymous_variant&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;low&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;synonymous_variant&quot;}</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;OR4F5&quot;}</td></tr>\n",
       "<tr><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;chr1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">358121</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;C&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;T&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1-358121-C-T&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;NC_000001.11:g.358121C&gt;T&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;SNV&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">358121</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">358121</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;PASS&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,60e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;1p36.33&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">2,00e-01</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">6,80e-02</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">4</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;ENST00000441866.2&quot;,&quot;Ensembl&quot;,&quot;lncRNA&quot;,&quot;ENSG00000228463&quot;,&quot;ENSG00000228463&quot;,[&quot;non_coding_transcript_exon_variant&quot;],&quot;modifier&quot;,False),(&quot;ENST00000450983.1&quot;,&quot;Ensembl&quot;,&quot;lncRNA&quot;,&quot;ENSG00000236601&quot;,&quot;ENSG00000236601&quot;,[&quot;upstream_gene_variant&quot;],&quot;modifier&quot;,True),(&quot;ENST00000412666.1&quot;,&quot;Ensembl&quot;,&quot;lncRNA&quot;,&quot;ENSG00000236601&quot;,&quot;ENSG00000236601&quot;,[&quot;upstream_gene_variant&quot;],&quot;modifier&quot;,False),(&quot;ENST00000635159.1&quot;,&quot;Ensembl&quot;,&quot;lncRNA&quot;,&quot;ENSG00000236601&quot;,&quot;ENSG00000236601&quot;,[&quot;upstream_gene_variant&quot;],&quot;modifier&quot;,False)]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[(&quot;1/1&quot;,[7,31e-01],93,3,[25,68])]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">chr1:358121</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;C&quot;,&quot;T&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">NA</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENST00000450983.1&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;Ensembl&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;lncRNA&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000236601&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;ENSG00000236601&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">[&quot;upstream_gene_variant&quot;]</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">&quot;modifier&quot;</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">True</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;non_coding_transcript_exon_variant&quot;,&quot;upstream_gene_variant&quot;}</td><td style=\"white-space: nowrap; max-width: 500px; overflow: hidden; text-overflow: ellipsis; \">{&quot;ENSG00000228463&quot;,&quot;ENSG00000236601&quot;}</td></tr>\n",
       "</tbody></table><p style=\"background: #fdd; padding: 0.4em;\">showing top 5 rows</p>\n"
      ],
      "text/plain": [
       "+------------+----------+-----+-----+----------------+\n",
       "| chromosome | position | ref | alt | vid            |\n",
       "+------------+----------+-----+-----+----------------+\n",
       "| str        |    int32 | str | str | str            |\n",
       "+------------+----------+-----+-----+----------------+\n",
       "| \"chr1\"     |    69081 | \"G\" | \"C\" | \"1-69081-G-C\"  |\n",
       "| \"chr1\"     |    69270 | \"A\" | \"G\" | \"1-69270-A-G\"  |\n",
       "| \"chr1\"     |    69511 | \"A\" | \"G\" | \"1-69511-A-G\"  |\n",
       "| \"chr1\"     |    69897 | \"T\" | \"C\" | \"1-69897-T-C\"  |\n",
       "| \"chr1\"     |   358121 | \"C\" | \"T\" | \"1-358121-C-T\" |\n",
       "+------------+----------+-----+-----+----------------+\n",
       "\n",
       "+----------------------------+--------------+--------+--------+---------+\n",
       "| hgvsg                      | variant_type |  begin |    end | filters |\n",
       "+----------------------------+--------------+--------+--------+---------+\n",
       "| str                        | str          |  int32 |  int32 | str     |\n",
       "+----------------------------+--------------+--------+--------+---------+\n",
       "| \"NC_000001.11:g.69081G>C\"  | \"SNV\"        |  69081 |  69081 | \"PASS\"  |\n",
       "| \"NC_000001.11:g.69270A>G\"  | \"SNV\"        |  69270 |  69270 | \"PASS\"  |\n",
       "| \"NC_000001.11:g.69511A>G\"  | \"SNV\"        |  69511 |  69511 | \"PASS\"  |\n",
       "| \"NC_000001.11:g.69897T>C\"  | \"SNV\"        |  69897 |  69897 | \"PASS\"  |\n",
       "| \"NC_000001.11:g.358121C>T\" | \"SNV\"        | 358121 | 358121 | \"PASS\"  |\n",
       "+----------------------------+--------------+--------+--------+---------+\n",
       "\n",
       "+-----------------+--------------------+---------+------------------+\n",
       "| mapping_quality | fisher_strand_bias | quality | cytogenetic_band |\n",
       "+-----------------+--------------------+---------+------------------+\n",
       "|         float64 |            float64 | float64 | str              |\n",
       "+-----------------+--------------------+---------+------------------+\n",
       "|        1,29e+01 |                 NA |      NA | \"1p36.33\"        |\n",
       "|        2,22e+01 |                 NA |      NA | \"1p36.33\"        |\n",
       "|        2,98e+01 |                 NA |      NA | \"1p36.33\"        |\n",
       "|        2,03e+01 |                 NA |      NA | \"1p36.33\"        |\n",
       "|        6,60e-01 |                 NA |      NA | \"1p36.33\"        |\n",
       "+-----------------+--------------------+---------+------------------+\n",
       "\n",
       "+--------------+----------------------+------------+------------+------+\n",
       "| phylop_score | phylop_primate_score | gerp_score | dann_score | rsid |\n",
       "+--------------+----------------------+------------+------------+------+\n",
       "|      float64 |              float64 |    float64 |    float64 | str  |\n",
       "+--------------+----------------------+------------+------------+------+\n",
       "|    -5,00e-01 |            -1,65e+00 |  -5,68e+00 |   3,60e-01 | NA   |\n",
       "|     8,00e-01 |             1,08e+00 |  -2,89e+00 |   5,80e-01 | NA   |\n",
       "|    -1,00e-01 |             6,99e-01 |   1,29e+00 |   4,50e-01 | NA   |\n",
       "|     7,00e-01 |             1,05e+00 |   9,04e-02 |   6,20e-01 | NA   |\n",
       "|     2,00e-01 |             6,80e-02 |         NA |         NA | NA   |\n",
       "+--------------+----------------------+------------+------------+------+\n",
       "\n",
       "+-----------+-----------+-----------+-----------+---------------+\n",
       "| gnomad_af | gnomad_ac | gnomad_an | gnomad_hc | gnomad_afr_af |\n",
       "+-----------+-----------+-----------+-----------+---------------+\n",
       "|   float64 |     int32 |     int32 |     int32 |       float64 |\n",
       "+-----------+-----------+-----------+-----------+---------------+\n",
       "|  8,50e-03 |       186 |     21896 |         5 |      2,01e-03 |\n",
       "|  6,29e-01 |     13469 |     21428 |      4877 |      3,54e-01 |\n",
       "|  8,46e-01 |     70558 |     83430 |     32520 |      5,95e-01 |\n",
       "|  4,86e-01 |     25259 |     51958 |      5492 |      2,92e-01 |\n",
       "|        NA |        NA |        NA |        NA |            NA |\n",
       "+-----------+-----------+-----------+-----------+---------------+\n",
       "\n",
       "+---------------+---------------+---------------+---------------+\n",
       "| gnomad_amr_af | gnomad_eas_af | gnomad_fin_af | gnomad_nfe_af |\n",
       "+---------------+---------------+---------------+---------------+\n",
       "|       float64 |       float64 |       float64 |       float64 |\n",
       "+---------------+---------------+---------------+---------------+\n",
       "|      1,32e-02 |      0,00e+00 |      3,10e-02 |      2,12e-02 |\n",
       "|      6,95e-01 |      9,96e-01 |      9,29e-01 |      8,80e-01 |\n",
       "|      8,95e-01 |      1,00e+00 |      9,91e-01 |      9,67e-01 |\n",
       "|      4,78e-01 |      6,04e-01 |      6,99e-01 |      6,75e-01 |\n",
       "|            NA |            NA |            NA |            NA |\n",
       "+---------------+---------------+---------------+---------------+\n",
       "\n",
       "+---------------+---------------+---------------+----------------------+\n",
       "| gnomad_asj_af | gnomad_sas_af | gnomad_oth_af | gnomad_failed_filter |\n",
       "+---------------+---------------+---------------+----------------------+\n",
       "|       float64 |       float64 |       float64 |                 bool |\n",
       "+---------------+---------------+---------------+----------------------+\n",
       "|      2,27e-02 |      5,17e-03 |            NA |                 True |\n",
       "|      8,03e-01 |      8,93e-01 |            NA |                 True |\n",
       "|      9,78e-01 |      9,78e-01 |            NA |                   NA |\n",
       "|      6,71e-01 |      6,26e-01 |            NA |                 True |\n",
       "|            NA |            NA |            NA |                   NA |\n",
       "+---------------+---------------+---------------+----------------------+\n",
       "\n",
       "+-----------------+-----------------+-----------------+-----------------+\n",
       "| gnomad_exome_af | gnomad_exome_ac | gnomad_exome_an | gnomad_exome_hc |\n",
       "+-----------------+-----------------+-----------------+-----------------+\n",
       "|         float64 |           int32 |           int32 |           int32 |\n",
       "+-----------------+-----------------+-----------------+-----------------+\n",
       "|        1,76e-02 |            4788 |          271898 |             729 |\n",
       "|        8,78e-01 |          374631 |          426620 |          176537 |\n",
       "|        9,62e-01 |          832303 |          865570 |          410006 |\n",
       "|        7,04e-01 |          209373 |          297420 |           78375 |\n",
       "|              NA |              NA |              NA |              NA |\n",
       "+-----------------+-----------------+-----------------+-----------------+\n",
       "\n",
       "+----------------------------+-----------+-----------+-----------+-----------+\n",
       "| gnomad_exome_failed_filter | topmed_af | topmed_ac | topmed_an | topmed_hc |\n",
       "+----------------------------+-----------+-----------+-----------+-----------+\n",
       "|                       bool |   float64 |     int32 |     int32 |     int32 |\n",
       "+----------------------------+-----------+-----------+-----------+-----------+\n",
       "|                       True |  8,29e-03 |      1041 |    125568 |         0 |\n",
       "|                       True |  2,75e-01 |     34480 |    125568 |        11 |\n",
       "|                       True |  8,75e-01 |    109911 |    125568 |     50718 |\n",
       "|                       True |  2,80e-01 |     35215 |    125568 |        25 |\n",
       "|                         NA |        NA |        NA |        NA |        NA |\n",
       "+----------------------------+-----------+-----------+-----------+-----------+\n",
       "\n",
       "+----------------------+----------------------+------------+---------------+\n",
       "| topmed_failed_filter | clinvar_significance | clinvar_id | n_transcripts |\n",
       "+----------------------+----------------------+------------+---------------+\n",
       "|                 bool | str                  | str        |         int32 |\n",
       "+----------------------+----------------------+------------+---------------+\n",
       "|                 True | NA                   | NA         |             3 |\n",
       "|                 True | NA                   | NA         |             2 |\n",
       "|                 True | NA                   | NA         |             2 |\n",
       "|                 True | NA                   | NA         |             2 |\n",
       "|                   NA | NA                   | NA         |             4 |\n",
       "+----------------------+----------------------+------------+---------------+\n",
       "\n",
       "+------------------------------------------------------------------------------+\n",
       "| transcripts                                                                  |\n",
       "+------------------------------------------------------------------------------+\n",
       "| array<struct{transcript_id: str, source: str, bio_type: str, gene_id: str... |\n",
       "+------------------------------------------------------------------------------+\n",
       "| [(\"ENST00000642116.1\",\"Ensembl\",\"lncRNA\",\"ENSG00000290826\",\"ENSG000002908... |\n",
       "| [(\"ENST00000641515.2\",\"Ensembl\",\"mRNA\",\"ENSG00000186092\",\"OR4F5\",[\"synony... |\n",
       "| [(\"ENST00000641515.2\",\"Ensembl\",\"mRNA\",\"ENSG00000186092\",\"OR4F5\",[\"missen... |\n",
       "| [(\"ENST00000641515.2\",\"Ensembl\",\"mRNA\",\"ENSG00000186092\",\"OR4F5\",[\"synony... |\n",
       "| [(\"ENST00000441866.2\",\"Ensembl\",\"lncRNA\",\"ENSG00000228463\",\"ENSG000002284... |\n",
       "+------------------------------------------------------------------------------+\n",
       "\n",
       "+------------------------------------------------------------------------------+\n",
       "| samples                                                                      |\n",
       "+------------------------------------------------------------------------------+\n",
       "| array<struct{genotype: str, variant_frequencies: array<float64>, total_de... |\n",
       "+------------------------------------------------------------------------------+\n",
       "| [(\"0/1\",[2,85e-01],116,9,[83,33])]                                           |\n",
       "| [(\"0/1\",[6,75e-01],83,3,[27,56])]                                            |\n",
       "| [(\"1/1\",[1,00e+00],68,41,[0,68])]                                            |\n",
       "| [(\"0/1\",[6,67e-01],102,6,[34,68])]                                           |\n",
       "| [(\"1/1\",[7,31e-01],93,3,[25,68])]                                            |\n",
       "+------------------------------------------------------------------------------+\n",
       "\n",
       "+---------------+------------+---------------+------------+\n",
       "| locus         | alleles    | max_gnomad_af | max_pop_af |\n",
       "+---------------+------------+---------------+------------+\n",
       "| locus<GRCh38> | array<str> |       float64 |    float64 |\n",
       "+---------------+------------+---------------+------------+\n",
       "| chr1:69081    | [\"G\",\"C\"]  |      1,76e-02 |   3,10e-02 |\n",
       "| chr1:69270    | [\"A\",\"G\"]  |      8,78e-01 |   9,96e-01 |\n",
       "| chr1:69511    | [\"A\",\"G\"]  |      9,62e-01 |   1,00e+00 |\n",
       "| chr1:69897    | [\"T\",\"C\"]  |      7,04e-01 |   6,99e-01 |\n",
       "| chr1:358121   | [\"C\",\"T\"]  |            NA |         NA |\n",
       "+---------------+------------+---------------+------------+\n",
       "\n",
       "+------------------------------------+-----------------------------+\n",
       "| canonical_transcript.transcript_id | canonical_transcript.source |\n",
       "+------------------------------------+-----------------------------+\n",
       "| str                                | str                         |\n",
       "+------------------------------------+-----------------------------+\n",
       "| \"ENST00000642116.1\"                | \"Ensembl\"                   |\n",
       "| \"ENST00000641515.2\"                | \"Ensembl\"                   |\n",
       "| \"ENST00000641515.2\"                | \"Ensembl\"                   |\n",
       "| \"ENST00000641515.2\"                | \"Ensembl\"                   |\n",
       "| \"ENST00000450983.1\"                | \"Ensembl\"                   |\n",
       "+------------------------------------+-----------------------------+\n",
       "\n",
       "+-------------------------------+------------------------------+\n",
       "| canonical_transcript.bio_type | canonical_transcript.gene_id |\n",
       "+-------------------------------+------------------------------+\n",
       "| str                           | str                          |\n",
       "+-------------------------------+------------------------------+\n",
       "| \"lncRNA\"                      | \"ENSG00000290826\"            |\n",
       "| \"mRNA\"                        | \"ENSG00000186092\"            |\n",
       "| \"mRNA\"                        | \"ENSG00000186092\"            |\n",
       "| \"mRNA\"                        | \"ENSG00000186092\"            |\n",
       "| \"lncRNA\"                      | \"ENSG00000236601\"            |\n",
       "+-------------------------------+------------------------------+\n",
       "\n",
       "+---------------------------+-----------------------------------+\n",
       "| canonical_transcript.hgnc | canonical_transcript.consequences |\n",
       "+---------------------------+-----------------------------------+\n",
       "| str                       | array<str>                        |\n",
       "+---------------------------+-----------------------------------+\n",
       "| \"ENSG00000290826\"         | [\"downstream_gene_variant\"]       |\n",
       "| \"OR4F5\"                   | [\"synonymous_variant\"]            |\n",
       "| \"OR4F5\"                   | [\"missense_variant\"]              |\n",
       "| \"OR4F5\"                   | [\"synonymous_variant\"]            |\n",
       "| \"ENSG00000236601\"         | [\"upstream_gene_variant\"]         |\n",
       "+---------------------------+-----------------------------------+\n",
       "\n",
       "+-----------------------------+-----------------------------------+\n",
       "| canonical_transcript.impact | canonical_transcript.is_canonical |\n",
       "+-----------------------------+-----------------------------------+\n",
       "| str                         |                              bool |\n",
       "+-----------------------------+-----------------------------------+\n",
       "| \"modifier\"                  |                              True |\n",
       "| \"low\"                       |                              True |\n",
       "| \"moderate\"                  |                              True |\n",
       "| \"low\"                       |                              True |\n",
       "| \"modifier\"                  |                              True |\n",
       "+-----------------------------+-----------------------------------+\n",
       "\n",
       "+----------------------------------------------------------------+\n",
       "| all_consequences                                               |\n",
       "+----------------------------------------------------------------+\n",
       "| set<str>                                                       |\n",
       "+----------------------------------------------------------------+\n",
       "| {\"downstream_gene_variant\",\"synonymous_variant\"}               |\n",
       "| {\"synonymous_variant\"}                                         |\n",
       "| {\"missense_variant\"}                                           |\n",
       "| {\"synonymous_variant\"}                                         |\n",
       "| {\"non_coding_transcript_exon_variant\",\"upstream_gene_variant\"} |\n",
       "+----------------------------------------------------------------+\n",
       "\n",
       "+---------------------------------------+\n",
       "| genes                                 |\n",
       "+---------------------------------------+\n",
       "| set<str>                              |\n",
       "+---------------------------------------+\n",
       "| {\"ENSG00000290826\",\"OR4F5\"}           |\n",
       "| {\"OR4F5\"}                             |\n",
       "| {\"OR4F5\"}                             |\n",
       "| {\"OR4F5\"}                             |\n",
       "| {\"ENSG00000228463\",\"ENSG00000236601\"} |\n",
       "+---------------------------------------+\n",
       "showing top 5 rows"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------------------------------------\n",
      "Global fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'chromosome': str \n",
      "    'position': int32 \n",
      "    'ref': str \n",
      "    'alt': str \n",
      "    'vid': str \n",
      "    'hgvsg': str \n",
      "    'variant_type': str \n",
      "    'begin': int32 \n",
      "    'end': int32 \n",
      "    'filters': str \n",
      "    'mapping_quality': float64 \n",
      "    'fisher_strand_bias': float64 \n",
      "    'quality': float64 \n",
      "    'cytogenetic_band': str \n",
      "    'phylop_score': float64 \n",
      "    'phylop_primate_score': float64 \n",
      "    'gerp_score': float64 \n",
      "    'dann_score': float64 \n",
      "    'rsid': str \n",
      "    'gnomad_af': float64 \n",
      "    'gnomad_ac': int32 \n",
      "    'gnomad_an': int32 \n",
      "    'gnomad_hc': int32 \n",
      "    'gnomad_afr_af': float64 \n",
      "    'gnomad_amr_af': float64 \n",
      "    'gnomad_eas_af': float64 \n",
      "    'gnomad_fin_af': float64 \n",
      "    'gnomad_nfe_af': float64 \n",
      "    'gnomad_asj_af': float64 \n",
      "    'gnomad_sas_af': float64 \n",
      "    'gnomad_oth_af': float64 \n",
      "    'gnomad_failed_filter': bool \n",
      "    'gnomad_exome_af': float64 \n",
      "    'gnomad_exome_ac': int32 \n",
      "    'gnomad_exome_an': int32 \n",
      "    'gnomad_exome_hc': int32 \n",
      "    'gnomad_exome_failed_filter': bool \n",
      "    'topmed_af': float64 \n",
      "    'topmed_ac': int32 \n",
      "    'topmed_an': int32 \n",
      "    'topmed_hc': int32 \n",
      "    'topmed_failed_filter': bool \n",
      "    'clinvar_significance': str \n",
      "    'clinvar_id': str \n",
      "    'n_transcripts': int32 \n",
      "    'transcripts': array<struct {\n",
      "        transcript_id: str, \n",
      "        source: str, \n",
      "        bio_type: str, \n",
      "        gene_id: str, \n",
      "        hgnc: str, \n",
      "        consequences: array<str>, \n",
      "        impact: str, \n",
      "        is_canonical: bool\n",
      "    }> \n",
      "    'samples': array<struct {\n",
      "        genotype: str, \n",
      "        variant_frequencies: array<float64>, \n",
      "        total_depth: int32, \n",
      "        genotype_quality: int32, \n",
      "        allele_depths: array<int32>\n",
      "    }> \n",
      "    'locus': locus<GRCh38> \n",
      "    'alleles': array<str> \n",
      "    'max_gnomad_af': float64 \n",
      "    'max_pop_af': float64 \n",
      "    'canonical_transcript': struct {\n",
      "        transcript_id: str, \n",
      "        source: str, \n",
      "        bio_type: str, \n",
      "        gene_id: str, \n",
      "        hgnc: str, \n",
      "        consequences: array<str>, \n",
      "        impact: str, \n",
      "        is_canonical: bool\n",
      "    } \n",
      "    'all_consequences': set<str> \n",
      "    'genes': set<str> \n",
      "----------------------------------------\n",
      "Key: ['locus', 'alleles']\n",
      "----------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Quick checks\n",
    "print(f\"Total variants: {ht.count()}\")\n",
    "print(f\"Chromosomes: {ht.aggregate(hl.agg.collect_as_set(ht.chromosome))}\")\n",
    "print(f\"Fields: {list(ht.row)}\")\n",
    "\n",
    "# Look at first few rows\n",
    "ht.show(5)\n",
    "\n",
    "# Get summary statistics\n",
    "ht.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9c181f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def explore_hail_table(ht: hl.Table):\n",
    "    \"\"\"\n",
    "    Comprehensive exploration and testing of a Hail Table\n",
    "    \n",
    "    Args:\n",
    "        ht (hl.Table): Hail Table to explore\n",
    "    \"\"\"\n",
    "    print(\"=\"*80)\n",
    "    print(\"HAIL TABLE EXPLORATION AND TESTING\")\n",
    "    print(\"=\"*80)\n",
    "    \n",
    "    # Basic info\n",
    "    print(\"\\n2. BASIC STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    n_variants = ht.count()\n",
    "    print(f\"   Total variants: {n_variants:,}\")\n",
    "    print(f\"   Key fields: {ht.key}\")\n",
    "    \n",
    "    # Schema\n",
    "    print(\"\\n3. SCHEMA OVERVIEW\")\n",
    "    print(\"-\" * 80)\n",
    "    ht.describe()\n",
    "    \n",
    "    # Field names\n",
    "    print(\"\\n4. ALL FIELD NAMES\")\n",
    "    print(\"-\" * 80)\n",
    "    fields = list(ht.row)\n",
    "    for i, field in enumerate(fields, 1):\n",
    "        print(f\"   {i:2d}. {field}\")\n",
    "    \n",
    "    # Sample data - first 5 rows\n",
    "    print(\"\\n5. SAMPLE DATA (First 5 variants)\")\n",
    "    print(\"-\" * 80)\n",
    "    sample_df = ht.head(5).to_pandas()\n",
    "    print(sample_df[['chromosome', 'position', 'ref', 'alt', 'vid', 'rsid']].to_string())\n",
    "    \n",
    "    # Chromosome distribution\n",
    "    print(\"\\n6. CHROMOSOME DISTRIBUTION\")\n",
    "    print(\"-\" * 80)\n",
    "    chrom_counts = ht.group_by(ht.chromosome).aggregate(n=hl.agg.count()).collect()\n",
    "    chrom_df = pd.DataFrame(chrom_counts).sort_values('chromosome')\n",
    "    print(chrom_df.to_string(index=False))\n",
    "    \n",
    "    # Variant type distribution\n",
    "    print(\"\\n7. VARIANT TYPE DISTRIBUTION\")\n",
    "    print(\"-\" * 80)\n",
    "    var_types = ht.group_by(ht.variant_type).aggregate(n=hl.agg.count()).collect()\n",
    "    var_type_df = pd.DataFrame(var_types).sort_values('n', ascending=False)\n",
    "    print(var_type_df.to_string(index=False))\n",
    "    \n",
    "    # gnomAD frequency statistics\n",
    "    print(\"\\n8. GNOMAD ALLELE FREQUENCY STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    gnomad_stats = ht.aggregate(hl.struct(\n",
    "        total_variants=hl.agg.count(),\n",
    "        with_gnomad_af=hl.agg.count_where(hl.is_defined(ht.gnomad_af)),\n",
    "        with_gnomad_exome_af=hl.agg.count_where(hl.is_defined(ht.gnomad_exome_af)),\n",
    "        mean_gnomad_af=hl.agg.mean(ht.gnomad_af),\n",
    "        median_gnomad_af=hl.agg.approx_median(ht.gnomad_af),\n",
    "        max_gnomad_af=hl.agg.max(ht.gnomad_af),\n",
    "    ))\n",
    "    print(f\"   Total variants: {gnomad_stats.total_variants:,}\")\n",
    "    print(f\"   With gnomAD genome AF: {gnomad_stats.with_gnomad_af:,}\")\n",
    "    print(f\"   With gnomAD exome AF: {gnomad_stats.with_gnomad_exome_af:,}\")\n",
    "    print(f\"   Mean gnomAD AF: {gnomad_stats.mean_gnomad_af:.6f}\" if gnomad_stats.mean_gnomad_af else \"   Mean gnomAD AF: N/A\")\n",
    "    print(f\"   Median gnomAD AF: {gnomad_stats.median_gnomad_af:.6f}\" if gnomad_stats.median_gnomad_af else \"   Median gnomAD AF: N/A\")\n",
    "    print(f\"   Max gnomAD AF: {gnomad_stats.max_gnomad_af:.6f}\" if gnomad_stats.max_gnomad_af else \"   Max gnomAD AF: N/A\")\n",
    "    \n",
    "    # Frequency bins\n",
    "    print(\"\\n9. GNOMAD AF FREQUENCY BINS\")\n",
    "    print(\"-\" * 80)\n",
    "    freq_bins = ht.aggregate(hl.struct(\n",
    "        rare_0_0001=hl.agg.count_where((ht.max_gnomad_af > 0) & (ht.max_gnomad_af <= 0.0001)),\n",
    "        rare_0_001=hl.agg.count_where((ht.max_gnomad_af > 0.0001) & (ht.max_gnomad_af <= 0.001)),\n",
    "        low_freq_0_01=hl.agg.count_where((ht.max_gnomad_af > 0.001) & (ht.max_gnomad_af <= 0.01)),\n",
    "        common_0_05=hl.agg.count_where((ht.max_gnomad_af > 0.01) & (ht.max_gnomad_af <= 0.05)),\n",
    "        very_common=hl.agg.count_where(ht.max_gnomad_af > 0.05),\n",
    "        no_data=hl.agg.count_where(~hl.is_defined(ht.max_gnomad_af)),\n",
    "    ))\n",
    "    print(f\"   Ultra-rare (AF  0.0001): {freq_bins.rare_0_0001:,}\")\n",
    "    print(f\"   Rare (0.0001 < AF  0.001): {freq_bins.rare_0_001:,}\")\n",
    "    print(f\"   Low freq (0.001 < AF  0.01): {freq_bins.low_freq_0_01:,}\")\n",
    "    print(f\"   Common (0.01 < AF  0.05): {freq_bins.common_0_05:,}\")\n",
    "    print(f\"   Very common (AF > 0.05): {freq_bins.very_common:,}\")\n",
    "    print(f\"   No gnomAD data: {freq_bins.no_data:,}\")\n",
    "    \n",
    "    # Transcript statistics\n",
    "    print(\"\\n10. TRANSCRIPT STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    transcript_stats = ht.aggregate(hl.struct(\n",
    "        with_transcripts=hl.agg.count_where(ht.n_transcripts > 0),\n",
    "        total_transcripts=hl.agg.sum(ht.n_transcripts),\n",
    "        with_canonical=hl.agg.count_where(hl.is_defined(ht.canonical_transcript)),\n",
    "        mean_transcripts=hl.agg.mean(ht.n_transcripts),\n",
    "        max_transcripts=hl.agg.max(ht.n_transcripts),\n",
    "    ))\n",
    "    print(f\"   Variants with transcripts: {transcript_stats.with_transcripts:,}\")\n",
    "    print(f\"   Total transcript annotations: {transcript_stats.total_transcripts:,}\")\n",
    "    print(f\"   Variants with canonical transcript: {transcript_stats.with_canonical:,}\")\n",
    "    print(f\"   Mean transcripts per variant: {transcript_stats.mean_transcripts:.2f}\")\n",
    "    print(f\"   Max transcripts per variant: {transcript_stats.max_transcripts}\")\n",
    "    \n",
    "    # ClinVar statistics\n",
    "    print(\"\\n11. CLINVAR ANNOTATIONS\")\n",
    "    print(\"-\" * 80)\n",
    "    clinvar_stats = ht.aggregate(hl.struct(\n",
    "        with_clinvar=hl.agg.count_where(hl.is_defined(ht.clinvar_significance)),\n",
    "        total=hl.agg.count(),\n",
    "    ))\n",
    "    print(f\"   Variants with ClinVar: {clinvar_stats.with_clinvar:,}\")\n",
    "    print(f\"   Percentage: {100 * clinvar_stats.with_clinvar / clinvar_stats.total:.2f}%\")\n",
    "    \n",
    "    # Conservation scores\n",
    "    print(\"\\n12. CONSERVATION SCORE STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    conservation_stats = ht.aggregate(hl.struct(\n",
    "        phylop_mean=hl.agg.mean(ht.phylop_score),\n",
    "        phylop_max=hl.agg.max(ht.phylop_score),\n",
    "        gerp_mean=hl.agg.mean(ht.gerp_score),\n",
    "        gerp_max=hl.agg.max(ht.gerp_score),\n",
    "    ))\n",
    "    print(f\"   PhyloP mean: {conservation_stats.phylop_mean:.3f}\" if conservation_stats.phylop_mean else \"   PhyloP mean: N/A\")\n",
    "    print(f\"   PhyloP max: {conservation_stats.phylop_max:.3f}\" if conservation_stats.phylop_max else \"   PhyloP max: N/A\")\n",
    "    print(f\"   GERP mean: {conservation_stats.gerp_mean:.3f}\" if conservation_stats.gerp_mean else \"   GERP mean: N/A\")\n",
    "    print(f\"   GERP max: {conservation_stats.gerp_max:.3f}\" if conservation_stats.gerp_max else \"   GERP max: N/A\")\n",
    "    \n",
    "    # Gene statistics\n",
    "    print(\"\\n13. GENE STATISTICS\")\n",
    "    print(\"-\" * 80)\n",
    "    gene_stats = ht.aggregate(hl.struct(\n",
    "        variants_with_genes=hl.agg.count_where(hl.len(ht.genes) > 0),\n",
    "    ))\n",
    "    # Count unique genes by converting set to array and aggregating\n",
    "    unique_genes_count = ht.aggregate(\n",
    "        hl.len(hl.agg.explode(lambda gene: hl.agg.collect_as_set(gene), hl.array(ht.genes)))\n",
    "    )\n",
    "    print(f\"   Variants with gene annotations: {gene_stats.variants_with_genes:,}\")\n",
    "    print(f\"   Unique genes: {unique_genes_count:,}\")\n",
    "    \n",
    "    # Top consequences\n",
    "    print(\"\\n14. TOP 10 CONSEQUENCES\")\n",
    "    print(\"-\" * 80)\n",
    "    consequences_ht = ht.explode(ht.all_consequences)\n",
    "    top_consequences = (consequences_ht\n",
    "                       .group_by(consequences_ht.all_consequences)\n",
    "                       .aggregate(n=hl.agg.count())\n",
    "                       .order_by(hl.desc('n'))\n",
    "                       .head(10)\n",
    "                       .collect())\n",
    "    for i, row in enumerate(top_consequences, 1):\n",
    "        print(f\"   {i:2d}. {row.all_consequences}: {row.n:,}\")\n",
    "    \n",
    "    # Example queries\n",
    "    print(\"\\n15. EXAMPLE VARIANT LOOKUPS\")\n",
    "    print(\"-\" * 80)\n",
    "    \n",
    "    # Get a few interesting variants\n",
    "    # Common variant\n",
    "    common = ht.filter(ht.max_gnomad_af > 0.1).head(1).collect()\n",
    "    if common:\n",
    "        v = common[0]\n",
    "        print(f\"\\n   Common variant (AF > 0.1):\")\n",
    "        print(f\"   {v.chromosome}:{v.position} {v.ref}>{v.alt}\")\n",
    "        print(f\"   rsID: {v.rsid}\")\n",
    "        print(f\"   gnomAD AF: {v.gnomad_af:.4f}\" if v.gnomad_af else \"   gnomAD AF: N/A\")\n",
    "        print(f\"   Genes: {', '.join(v.genes) if v.genes else 'N/A'}\")\n",
    "    \n",
    "    # Rare variant\n",
    "    rare = ht.filter((ht.max_gnomad_af > 0) & (ht.max_gnomad_af < 0.001)).head(1).collect()\n",
    "    if rare:\n",
    "        v = rare[0]\n",
    "        print(f\"\\n   Rare variant (AF < 0.001):\")\n",
    "        print(f\"   {v.chromosome}:{v.position} {v.ref}>{v.alt}\")\n",
    "        print(f\"   gnomAD AF: {v.gnomad_af:.6f}\" if v.gnomad_af else \"   gnomAD AF: N/A\")\n",
    "        print(f\"   Consequences: {', '.join(list(v.all_consequences)) if v.all_consequences else 'N/A'}\")\n",
    "    \n",
    "    # Variant with ClinVar\n",
    "    clinvar_var = ht.filter(hl.is_defined(ht.clinvar_significance)).head(1).collect()\n",
    "    if clinvar_var:\n",
    "        v = clinvar_var[0]\n",
    "        print(f\"\\n   ClinVar variant:\")\n",
    "        print(f\"   {v.chromosome}:{v.position} {v.ref}>{v.alt}\")\n",
    "        print(f\"   ClinVar significance: {v.clinvar_significance}\")\n",
    "        print(f\"   ClinVar ID: {v.clinvar_id}\")\n",
    "    \n",
    "    print(\"\\n\" + \"=\"*80)\n",
    "    print(\"TESTING COMPLETE!\")\n",
    "    print(\"=\"*80)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0e4d4ee1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "HAIL TABLE EXPLORATION AND TESTING\n",
      "================================================================================\n",
      "\n",
      "2. BASIC STATISTICS\n",
      "--------------------------------------------------------------------------------\n",
      "   Total variants: 30,487\n",
      "   Key fields: <StructExpression of type struct{locus: locus<GRCh38>, alleles: array<str>}>\n",
      "\n",
      "3. SCHEMA OVERVIEW\n",
      "--------------------------------------------------------------------------------\n",
      "----------------------------------------\n",
      "Global fields:\n",
      "    None\n",
      "----------------------------------------\n",
      "Row fields:\n",
      "    'chromosome': str \n",
      "    'position': int32 \n",
      "    'ref': str \n",
      "    'alt': str \n",
      "    'vid': str \n",
      "    'hgvsg': str \n",
      "    'variant_type': str \n",
      "    'begin': int32 \n",
      "    'end': int32 \n",
      "    'filters': str \n",
      "    'mapping_quality': float64 \n",
      "    'fisher_strand_bias': float64 \n",
      "    'quality': float64 \n",
      "    'cytogenetic_band': str \n",
      "    'phylop_score': float64 \n",
      "    'phylop_primate_score': float64 \n",
      "    'gerp_score': float64 \n",
      "    'dann_score': float64 \n",
      "    'rsid': str \n",
      "    'gnomad_af': float64 \n",
      "    'gnomad_ac': int32 \n",
      "    'gnomad_an': int32 \n",
      "    'gnomad_hc': int32 \n",
      "    'gnomad_afr_af': float64 \n",
      "    'gnomad_amr_af': float64 \n",
      "    'gnomad_eas_af': float64 \n",
      "    'gnomad_fin_af': float64 \n",
      "    'gnomad_nfe_af': float64 \n",
      "    'gnomad_asj_af': float64 \n",
      "    'gnomad_sas_af': float64 \n",
      "    'gnomad_oth_af': float64 \n",
      "    'gnomad_failed_filter': bool \n",
      "    'gnomad_exome_af': float64 \n",
      "    'gnomad_exome_ac': int32 \n",
      "    'gnomad_exome_an': int32 \n",
      "    'gnomad_exome_hc': int32 \n",
      "    'gnomad_exome_failed_filter': bool \n",
      "    'topmed_af': float64 \n",
      "    'topmed_ac': int32 \n",
      "    'topmed_an': int32 \n",
      "    'topmed_hc': int32 \n",
      "    'topmed_failed_filter': bool \n",
      "    'clinvar_significance': str \n",
      "    'clinvar_id': str \n",
      "    'n_transcripts': int32 \n",
      "    'transcripts': array<struct {\n",
      "        transcript_id: str, \n",
      "        source: str, \n",
      "        bio_type: str, \n",
      "        gene_id: str, \n",
      "        hgnc: str, \n",
      "        consequences: array<str>, \n",
      "        impact: str, \n",
      "        is_canonical: bool\n",
      "    }> \n",
      "    'samples': array<struct {\n",
      "        genotype: str, \n",
      "        variant_frequencies: array<float64>, \n",
      "        total_depth: int32, \n",
      "        genotype_quality: int32, \n",
      "        allele_depths: array<int32>\n",
      "    }> \n",
      "    'locus': locus<GRCh38> \n",
      "    'alleles': array<str> \n",
      "    'max_gnomad_af': float64 \n",
      "    'max_pop_af': float64 \n",
      "    'canonical_transcript': struct {\n",
      "        transcript_id: str, \n",
      "        source: str, \n",
      "        bio_type: str, \n",
      "        gene_id: str, \n",
      "        hgnc: str, \n",
      "        consequences: array<str>, \n",
      "        impact: str, \n",
      "        is_canonical: bool\n",
      "    } \n",
      "    'all_consequences': set<str> \n",
      "    'genes': set<str> \n",
      "----------------------------------------\n",
      "Key: ['locus', 'alleles']\n",
      "----------------------------------------\n",
      "\n",
      "4. ALL FIELD NAMES\n",
      "--------------------------------------------------------------------------------\n",
      "    1. chromosome\n",
      "    2. position\n",
      "    3. ref\n",
      "    4. alt\n",
      "    5. vid\n",
      "    6. hgvsg\n",
      "    7. variant_type\n",
      "    8. begin\n",
      "    9. end\n",
      "   10. filters\n",
      "   11. mapping_quality\n",
      "   12. fisher_strand_bias\n",
      "   13. quality\n",
      "   14. cytogenetic_band\n",
      "   15. phylop_score\n",
      "   16. phylop_primate_score\n",
      "   17. gerp_score\n",
      "   18. dann_score\n",
      "   19. rsid\n",
      "   20. gnomad_af\n",
      "   21. gnomad_ac\n",
      "   22. gnomad_an\n",
      "   23. gnomad_hc\n",
      "   24. gnomad_afr_af\n",
      "   25. gnomad_amr_af\n",
      "   26. gnomad_eas_af\n",
      "   27. gnomad_fin_af\n",
      "   28. gnomad_nfe_af\n",
      "   29. gnomad_asj_af\n",
      "   30. gnomad_sas_af\n",
      "   31. gnomad_oth_af\n",
      "   32. gnomad_failed_filter\n",
      "   33. gnomad_exome_af\n",
      "   34. gnomad_exome_ac\n",
      "   35. gnomad_exome_an\n",
      "   36. gnomad_exome_hc\n",
      "   37. gnomad_exome_failed_filter\n",
      "   38. topmed_af\n",
      "   39. topmed_ac\n",
      "   40. topmed_an\n",
      "   41. topmed_hc\n",
      "   42. topmed_failed_filter\n",
      "   43. clinvar_significance\n",
      "   44. clinvar_id\n",
      "   45. n_transcripts\n",
      "   46. transcripts\n",
      "   47. samples\n",
      "   48. locus\n",
      "   49. alleles\n",
      "   50. max_gnomad_af\n",
      "   51. max_pop_af\n",
      "   52. canonical_transcript\n",
      "   53. all_consequences\n",
      "   54. genes\n",
      "\n",
      "5. SAMPLE DATA (First 5 variants)\n",
      "--------------------------------------------------------------------------------\n",
      "  chromosome  position ref alt           vid  rsid\n",
      "0       chr1     69081   G   C   1-69081-G-C  <NA>\n",
      "1       chr1     69270   A   G   1-69270-A-G  <NA>\n",
      "2       chr1     69511   A   G   1-69511-A-G  <NA>\n",
      "3       chr1     69897   T   C   1-69897-T-C  <NA>\n",
      "4       chr1    358121   C   T  1-358121-C-T  <NA>\n",
      "\n",
      "6. CHROMOSOME DISTRIBUTION\n",
      "--------------------------------------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 5:==================================>                     (51 + 13) / 84]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome    n\n",
      "      chr1 3036\n",
      "     chr10 1168\n",
      "     chr11 2140\n",
      "     chr12 1460\n",
      "     chr13  605\n",
      "     chr14 1105\n",
      "     chr15 1008\n",
      "     chr16 1348\n",
      "     chr17 1810\n",
      "     chr18  501\n",
      "     chr19 2453\n",
      "      chr2 2040\n",
      "     chr20  698\n",
      "     chr21  418\n",
      "     chr22  811\n",
      "      chr3 1433\n",
      "      chr4 1203\n",
      "      chr5 1373\n",
      "      chr6 1669\n",
      "      chr7 1507\n",
      "      chr8 1037\n",
      "      chr9 1244\n",
      "      chrX  417\n",
      "      chrY    3\n",
      "\n",
      "7. VARIANT TYPE DISTRIBUTION\n",
      "--------------------------------------------------------------------------------\n",
      "    n variant_type\n",
      "29008          SNV\n",
      "  759     deletion\n",
      "  718    insertion\n",
      "    1          MNV\n",
      "    1       delins\n",
      "\n",
      "8. GNOMAD ALLELE FREQUENCY STATISTICS\n",
      "--------------------------------------------------------------------------------\n",
      "   Total variants: 30,487\n",
      "   With gnomAD genome AF: 29,915\n",
      "   With gnomAD exome AF: 28,146\n",
      "   Mean gnomAD AF: 0.447809\n",
      "   Median gnomAD AF: 0.422766\n",
      "   Max gnomAD AF: 1.000000\n",
      "\n",
      "9. GNOMAD AF FREQUENCY BINS\n",
      "--------------------------------------------------------------------------------\n",
      "   Ultra-rare (AF  0.0001): 504\n",
      "   Rare (0.0001 < AF  0.001): 346\n",
      "   Low freq (0.001 < AF  0.01): 713\n",
      "   Common (0.01 < AF  0.05): 1,267\n",
      "   Very common (AF > 0.05): 27,152\n",
      "   No gnomAD data: 349\n",
      "\n",
      "10. TRANSCRIPT STATISTICS\n",
      "--------------------------------------------------------------------------------\n",
      "   Variants with transcripts: 30,362\n",
      "   Total transcript annotations: 388,623\n",
      "   Variants with canonical transcript: 30,201\n",
      "   Mean transcripts per variant: 12.75\n",
      "   Max transcripts per variant: 402\n",
      "\n",
      "11. CLINVAR ANNOTATIONS\n",
      "--------------------------------------------------------------------------------\n",
      "   Variants with ClinVar: 9,154\n",
      "   Percentage: 30.03%\n",
      "\n",
      "12. CONSERVATION SCORE STATISTICS\n",
      "--------------------------------------------------------------------------------\n",
      "   PhyloP mean: -0.527\n",
      "   PhyloP max: 1.200\n",
      "   GERP mean: -2.095\n",
      "   GERP max: 8.050\n",
      "\n",
      "13. GENE STATISTICS\n",
      "--------------------------------------------------------------------------------\n",
      "   Variants with gene annotations: 30,362\n",
      "   Unique genes: 17,458\n",
      "\n",
      "14. TOP 10 CONSEQUENCES\n",
      "--------------------------------------------------------------------------------\n",
      "    1. downstream_gene_variant: 13,197\n",
      "    2. missense_variant: 12,800\n",
      "    3. non_coding_transcript_exon_variant: 12,784\n",
      "    4. synonymous_variant: 12,754\n",
      "    5. intron_variant: 12,161\n",
      "    6. upstream_gene_variant: 11,418\n",
      "    7. non_coding_transcript_variant: 7,330\n",
      "    8. 3_prime_UTR_variant: 5,627\n",
      "    9. 5_prime_UTR_variant: 1,681\n",
      "   10. splice_region_variant: 846\n",
      "\n",
      "15. EXAMPLE VARIANT LOOKUPS\n",
      "--------------------------------------------------------------------------------\n",
      "\n",
      "   Common variant (AF > 0.1):\n",
      "   chr1:69270 A>G\n",
      "   rsID: None\n",
      "   gnomAD AF: 0.6286\n",
      "   Genes: OR4F5\n",
      "\n",
      "   Rare variant (AF < 0.001):\n",
      "   chr1:1041200 C>T\n",
      "   gnomAD AF: 0.000013\n",
      "   Consequences: missense_variant, upstream_gene_variant, non_coding_transcript_exon_variant\n",
      "\n",
      "   ClinVar variant:\n",
      "   chr1:942451 T>C\n",
      "   ClinVar significance: benign;uncertain significance;uncertain significance;benign\n",
      "   ClinVar ID: VCV001166513.9;VCV003249251.1;RCV004814516.1;RCV001514847.8\n",
      "\n",
      "================================================================================\n",
      "TESTING COMPLETE!\n",
      "================================================================================\n"
     ]
    }
   ],
   "source": [
    "explore_hail_table(ht)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4405be8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_examples(ht: hl.Table):\n",
    "    \"\"\"\n",
    "    Examples of useful queries you can run\n",
    "    \n",
    "    Args:\n",
    "        ht (hl.Table): Hail Table to query\n",
    "    \"\"\"\n",
    "    print(\"\\n\" + \"=\"*80)\n",
    "    print(\"EXAMPLE QUERIES\")\n",
    "    print(\"=\"*80)\n",
    "    \n",
    "    # 1. Filter by frequency\n",
    "    print(\"\\n1. Get rare variants (AF < 0.01)\")\n",
    "    rare_variants = ht.filter(\n",
    "        (ht.max_gnomad_af > 0) & (ht.max_gnomad_af < 0.01)\n",
    "    )\n",
    "    print(f\"   Found {rare_variants.count():,} rare variants\")\n",
    "    \n",
    "    # 2. Filter by gene\n",
    "    print(\"\\n2. Get variants in specific gene (e.g., first gene in dataset)\")\n",
    "    first_gene = ht.filter(hl.len(ht.genes) > 0).head(1).collect()[0].genes\n",
    "    if first_gene:\n",
    "        gene_name = list(first_gene)[0]\n",
    "        gene_variants = ht.filter(ht.genes.contains(gene_name))\n",
    "        print(f\"   Gene: {gene_name}\")\n",
    "        print(f\"   Variants: {gene_variants.count()}\")\n",
    "    \n",
    "    # 3. Filter by consequence\n",
    "    print(\"\\n3. Get missense variants\")\n",
    "    missense = ht.filter(ht.all_consequences.contains('missense_variant'))\n",
    "    print(f\"   Found {missense.count():,} missense variants\")\n",
    "    \n",
    "    # 4. Filter by ClinVar pathogenic\n",
    "    print(\"\\n4. Get pathogenic variants\")\n",
    "    pathogenic = ht.filter(\n",
    "        hl.is_defined(ht.clinvar_significance) &\n",
    "        ht.clinvar_significance.contains('Pathogenic')\n",
    "    )\n",
    "    print(f\"   Found {pathogenic.count():,} pathogenic variants\")\n",
    "    \n",
    "    # 5. Filter by conservation\n",
    "    print(\"\\n5. Get highly conserved variants (PhyloP > 2)\")\n",
    "    conserved = ht.filter(ht.phylop_score > 2)\n",
    "    print(f\"   Found {conserved.count():,} highly conserved variants\")\n",
    "    \n",
    "    # 6. Complex filter\n",
    "    print(\"\\n6. Complex filter: Rare + missense + conserved\")\n",
    "    complex_filter = ht.filter(\n",
    "        (ht.max_gnomad_af < 0.01) &\n",
    "        ht.all_consequences.contains('missense_variant') &\n",
    "        (ht.phylop_score > 2)\n",
    "    )\n",
    "    print(f\"   Found {complex_filter.count():,} variants\")\n",
    "    \n",
    "    print(\"\\n\" + \"=\"*80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "efba17f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================================================================================\n",
      "EXAMPLE QUERIES\n",
      "================================================================================\n",
      "\n",
      "1. Get rare variants (AF < 0.01)\n",
      "   Found 1,563 rare variants\n",
      "\n",
      "2. Get variants in specific gene (e.g., first gene in dataset)\n",
      "   Gene: ENSG00000290826\n",
      "   Variants: 1\n",
      "\n",
      "3. Get missense variants\n",
      "   Found 12,800 missense variants\n",
      "\n",
      "4. Get pathogenic variants\n",
      "   Found 0 pathogenic variants\n",
      "\n",
      "5. Get highly conserved variants (PhyloP > 2)\n",
      "   Found 0 highly conserved variants\n",
      "\n",
      "6. Complex filter: Rare + missense + conserved\n",
      "   Found 0 variants\n",
      "\n",
      "================================================================================\n"
     ]
    }
   ],
   "source": [
    "query_examples(ht)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gnomad-toolbox",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
